agent/cache-types/resolved_service_config.go

```diff
@@ -0,0 +1,52 @@
+package cachetype
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/consul/agent/cache"
+	"github.com/hashicorp/consul/agent/structs"
+)
+
+// Recommended name for registration.
+const ResolvedServiceConfigName = "resolved-service-config"
+
+// ResolvedServiceConfig supports fetching the config for a service resolved from
+// the global proxy defaults and the centrally registered service config.
+type ResolvedServiceConfig struct {
+	RPC RPC
+}
+
+func (c *ResolvedServiceConfig) Fetch(opts cache.FetchOptions, req cache.Request) (cache.FetchResult, error) {
+	var result cache.FetchResult
+
+	// The request should be a ServiceConfigRequest.
+	reqReal, ok := req.(*structs.ServiceConfigRequest)
+	if !ok {
+		return result, fmt.Errorf(
+			"Internal cache failure: request wrong type: %T", req)
+	}
+
+	// Set the minimum query index to our current index so we block
+	reqReal.QueryOptions.MinQueryIndex = opts.MinIndex
+	reqReal.QueryOptions.MaxQueryTime = opts.Timeout
+
+	// Allways allow stale - there's no point in hitting leader if the request is
+	// going to be served from cache and endup arbitrarily stale anyway. This
+	// allows cached service-discover to automatically read scale across all
+	// servers too.
+	reqReal.AllowStale = true
+
+	// Fetch
+	var reply structs.ServiceConfigResponse
+	if err := c.RPC.RPC("ConfigEntry.ResolveServiceConfig", reqReal, &reply); err != nil {
+		return result, err
+	}
+
+	result.Value = &reply
+	result.Index = reply.QueryMeta.Index
+	return result, nil
+}
+
+func (c *ResolvedServiceConfig) SupportsBlocking() bool {
+	return true
+}
```

Comment:
Kinda sad this file is 100% boiler plate with only types changed from other cache types. I know @mkeeler felt strongly about that before. I kinda dislike it but not enough to refactor caching again or add bad abstractions!